packages/typescript-estree/tests/lib/persistentParse.ts

```diff
@@ -0,0 +1,136 @@
+import fs from 'fs';
+import path from 'path';
+import tmp from 'tmp';
+import { parseAndGenerateServices } from '../../src/parser';
+import { clearCaches } from '../../src/tsconfig-parser';
+
+const tsConfigExcludeBar = {
+  include: ['./*.ts'],
+  exclude: ['./bar.ts'],
+};
+const tsConfigIncludeAll = {
+  include: ['./*.ts'],
+  exclude: [],
+};
+const CONTENTS = {
+  foo: 'console.log("foo")',
+  bar: 'console.log("bar")',
+};
+
+const tmpDirs = new Set<tmp.DirResult>();
+afterEach(() => {
+  // stop watching the files and folders
+  clearCaches();
+
+  // clean up the temporary files and folders
+  tmpDirs.forEach(t => t.removeCallback());
+  tmpDirs.clear();
+});
+
+function writeTSConfig(
+  dirName: string,
+  config: Record<string, string[]>,
+): void {
+  fs.writeFileSync(path.join(dirName, 'tsconfig.json'), JSON.stringify(config));
+}
+function writeFile(dirName: string, file: 'foo' | 'bar'): void {
+  fs.writeFileSync(path.join(dirName, `${file}.ts`), CONTENTS[file]);
+}
+
+function setup(tsconfig: Record<string, string[]>, writeBar = true): string {
+  const tmpDir = tmp.dirSync({
```

Comment:
I didn't really want to do this test with real files on the filesystem, but it seemed too complicated to create my own mock filesystem. 

Jest uses `metro-memory-fs`, but this didn't support file watching out of the box, and I couldn't be bothered implementing it.